### ISSUE TITLE
Feature Request: OnRetryLimitReached

### ISSUE THREAD

**Issue by TheColonel2688** [z4kn4fein/trybot#3]
Could we have an OnRetyLimitExceeded() method? That is called after a retry is completed successfully?

I am using TryBot Retry together with Stateless (A StateMachine Library) and I need to be able to signal to the state machine that an all of the retries failed. 

I can use a FallBack and catch Trybot.Retry.Exceptions.MaxRetryAttemptsReachedException but it seems a little messy for my purpose.

Sample usage.

```
var policyForCommands = new BotPolicy();

policyForCommands.Configure(config => config
	.Retry(retryconfig => retryconfig
		.WhenExceptionOccurs(exception => exception is Exception)
		.WithMaxAttemptCount(5)
		.OnRetry((exception, context) =>
		{
			Console.WriteLine("Retrying ");
			//** StateMachine Change State To "ErrorOccured"
		})
		.OnRetryLimitExceeded((Context) => 
		{
			Console.WriteLine("All Retries Failed");
			//** StateMachine Change State To "TryingToReconnect"
		})
		)
);
```

**Comment by z4kn4fein** [comment]
Sure, i'll add this functionality soon!

**Comment by TheColonel2688** [comment]
It would be fine if after, OnRetryLimitExceeded, it throws the MaxRetryAttemptsReachedException.

HOWEVER: I think it would be even better if you could throw an exception at the end of the OnRetryLimitExceeded Method Block and have that thrown instead of  MaxRetryAttemptsReachedException. If no exception was thrown then the default MaxRetryAttemptsReachedException would be thrown. Not sure how easy or hard that would be to implement but it would be cool. 

I foresee using this a lot when I make client classes for network, USB, and serial devices.

And thank you

**Comment by z4kn4fein** [comment]
Cool, i'm happy to hear that!
I'm not sure i understand your suggestion correctly about the different exception, it's clear that it should be thrown after the OnRetryLimitExceeded but what kind of exception you'd like to see instead of the MaxRetryAttemptsReachedException?

**Comment by TheColonel2688** [comment]
Ok so this is a very crude example of implementing the features I have mentioned. I am sure implementing them inside TryBot is more involved but this should demonstrate basic behavior.

Basically the idea is that you can throw an exception inside the OnRetryLimitReached action/delegate and Trybot will catch that and throw it instead of the MaxRetryAttemptsReachedException.

If the user of Trybot felt so inclined they could still create a  MaxRetryAttemptsReachedException and add the exception from the user code as the inner exception and then throw a second user exception with MaxRetryAttemptsReachedException as the inner exception for that.  so you would have 2 nest exception inside the Exception that gets thrown back to the user.

UserExceptionB
    MaxRetryAttemptsReachedException
        UserExceptionA

This is more of an idea for you to think about. I am still debating myself whether it would be a good idea or not.

Any way here is my crude pseudo code example.

```
            Action userAction = () => { throw new InvalidOperationException("Demo/Mock Exception"); };

            Action<Exception> OnRetry = (exception) => { /* I failed, retrying */ };

            Action OnRetrySuccess = () => { /* Yaaaaay I succeeded, after I failed atleast once */ };

            Action<Exception> OnMaxRetryAttemptsReached = (userCodeException) => 
            {
                /* this block is defined by user */
                var IFeelLikeBeingVerbose = false;
                if (IFeelLikeBeingVerbose)
                {
                    var trybotException = new TryBot.MaxRetryAttemptsReachedException("Max Retry Attempts Reached", userCodeException);

                    throw new SomeUserException(" Some User Exception", trybotException);
                }
                else
                {
                    throw new SomeUserException(" Some User Exception", userCodeException);
                }

            };

            Func<Exception, bool> exceptionToLookFor = e => e is InvalidOperationException;

            int numberOfRetries = 5;

            //** in side trybot policy
            {
                int retryCount = 0;
                //for(int i = 0; i < numberOfRetries; i++)
                while(true)
                {
                    try
                    {
                        userAction.Invoke();

                        if(retryCount > 0)
                        {
                            OnRetrySuccess.Invoke();
                        }
                        break;
                    }
                    catch (Exception ex)
                    {
                        if (exceptionToLookFor.Invoke(ex))
                        {
                            if(retryCount++ < numberOfRetries)
                            {
                                OnRetry.Invoke(ex);
                                continue;
                            }
                            else
                            {
                                try
                                {
                                    OnMaxRetryAttemptsReached.Invoke(ex);
                                }
                                catch (Exception)
                                {
                                    //** Rethrow exception thrown in OnMaxRetry Do not throw TryBotMaxRetryAttemptsReachedException as defined below
                                    throw;
                                }

                                throw new TryBot.MaxRetryAttemptsReachedException("Max Retry Attempts Reached", ex);
                            }
 
                        }
                    }
                }
            }
```

**Comment by z4kn4fein** [comment]
This is still confusing for me a bit, i'm not sure that the framework should listen to user related exceptions thrown in the event handlers. 

My original idea was that when the user subscribed to the `OnMaxRetryAttemptsReached` event the framework shouldn't throw any exception at all but just fire the event. For me this indicates that the user wants to handle the limit reached case with a reaction to that event instead of catching exceptions. 
What do you think about this?

**Comment by TheColonel2688** [comment]
hmmm well like I said I'm still conflicted on it.

Does the MaxRetryAttemptsReachedException contain an inner exception with the exception that was throw from the users code? In my case there is only one reason for a call to fail, or at least I can generalize it to one reason. So I never actually looked. If it does contain an inner exception then that could be pulled out on a try catch around the call in to the policy. If not then throwing a new exception from the event handler would be another way to tell the caller why it failed.

What I do know is that my case I do need some kind of an exception thrown because I need to have something to tell the original caller that the call failed. The use for new event handler in my case is to also tell another task "flow" if you will (technically not another thread) that the call failed. This way it can execute code to recover.

The more I think about it. Listening for the exception in the event handler is not really needed, nor does it accomplish anything because I could still choose to throw an exception in the even handler and it would break out back to the caller correct? So if you chose not to have framework throw an exception in the case that the event handler is subscribed to, the user could still throw an exception in the handler after they were done executing other code.

To sum all this up. I need a way to tell my other task flow of the failure and tell the original caller that the call failed even if I subscribe to the MaxRetryAttenptsReached. Also for future use of the library it would be nice to know why. How you do it exactly I don't really have a strong opinion on.

Clear as mud right?

**Comment by z4kn4fein** [comment]
Yep, it contains the underlying exception thrown by the user code as its inner exception. After a second round of thinking it seems reasonable to also throw the `MaxRetryAttemptsReachedException` and fire the event simultaneously, and yes if you throw an exception in any of the callbacks that will bubble up to the place where you made the call on the policy.

So just a final clarification: the framework will invoke this event and then will throw a `MaxRetryAttemptsReachedException` right after each other, is this okay for you?

**Comment by TheColonel2688** [comment]
yup that works for me :)

**Comment by z4kn4fein** [comment]
Hey, i pushed this one also, could you please check this too?

**Comment by TheColonel2688** [comment]
yes will do

**Comment by TheColonel2688** [comment]
I tested it out, it seems to work like I would expect.

Thank you very much appreciate the quick feature implementation

**Comment by z4kn4fein** [comment]
released in [2.4.3](https://github.com/z4kn4fein/trybot/releases/tag/2.4.3)
